subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java

```diff
@@ -307,7 +322,46 @@ public void useJUnitJupiter(Provider<String> version) {
 
     @Override
     public void useSpock() {
-        useSpock(TestingFramework.SPOCK.getDefaultVersion());
+        useSpock(getDerivedSpockVersion());
+    }
+
+    private static final Set<String> GROOVY_GROUPS = new HashSet<>(Arrays.asList("org.codehaus.groovy", "org.apache.groovy"));
+    private static final Set<String> GROOVY_NAMES = new HashSet<>(Arrays.asList("groovy", "groovy-all"));
+    private static final Map<VersionNumber, String> GROOVY_TO_SPOCK_VERSION_MAPPING = getSpockVersionMapping(ImmutableMap.of(
+        "4.0", "2.2-groovy-4.0",
+        "3.0", "2.2-groovy-3.0",
+        "2.5", "2.2-groovy-2.5",
+        "2.4", "1.3-groovy-2.4",
+        "2.3", "1.1-groovy-2.3",
+        "2.0", "1.1-groovy-2.0",
+        "1.8", "0.7-groovy-1.8",
+        "1.7", "0.6-groovy-1.7",
+        "1.6", "0.5-groovy-1.6"
+    ));
+
+    private Provider<String> getDerivedSpockVersion() {
+        return getProviderFactory().provider(() -> {
+            Optional<VersionNumber> maybeGroovyJarVersionNumber = runtimeClasspath.getIncoming().getResolutionResult().getAllComponents().stream()
+                .map(ResolvedComponentResult::getModuleVersion)
+                .filter(mvi -> GROOVY_GROUPS.contains(mvi.getGroup()))
+                .filter(mvi -> GROOVY_NAMES.contains(mvi.getName()))
+                .map(ModuleVersionIdentifier::getVersion)
+                .map(VersionNumber::parse)
+                .findFirst();
+
+            // TODO: Should we warn if we can't find Groovy on their classpath?
+            VersionNumber groovyVersion = maybeGroovyJarVersionNumber.orElse(VersionNumber.parse(GroovySystem.getVersion()));
+
+            return GROOVY_TO_SPOCK_VERSION_MAPPING.entrySet().stream()
+                .filter(it -> groovyVersion.compareTo(it.getKey()) >= 0)
+                .map(Map.Entry::getValue)
+                .findFirst()
+                .orElse(TestingFramework.SPOCK.getDefaultVersion());
```

Comment:
I like this, especially the `findFirst().orElse(...)`.  My only concern is iteration order.  If the stream always handles entries in insertion order, we're good.  But I'm not sure if Guava's ImmutableMap implementation is a LinkedHashMap or not.

Reply:
It's not exactly a LinkedHashMap, but it does preserve order.

Reply:
Great. Maybe it's worth a comment on GROOVY_TO_SPOCK_VERSION_MAPPING that new future entries should be added in descending order?

Reply:
Easy to check the stream like this:

`entrySet().stream().spliterator().hasCharacteristics(Spliterator.SORTED)`

Reply:
You actually want https://docs.oracle.com/javase/8/docs/api/java/util/Spliterator.html#Ordered here.